cwl/src/main/scala/cwl/CwlWomExpression.scala

```diff
@@ -49,65 +50,160 @@ final case class InitialWorkDirFileGeneratorExpression(entry: IwdrListingArrayEn
+    val parameterContext = ParameterContext(inputValues)
+    entry.fold(InitialWorkDirFilePoly).apply(ioFunctionSet, parameterContext, expressionLib)
+  }
+
+  override def evaluateFiles(inputTypes: Map[String, WomValue],
+                             ioFunctionSet: IoFunctionSet,
+                             coerceTo: WomType): ErrorOr[Set[WomFile]] = {
+    "Programmer error: Shouldn't use InitialWorkDirRequirement listing to find output files.".invalidNel
+  }
+
+  /**
+    * We already get all of the task inputs when evaluating, and we don't need to highlight anything else
+    */
+  override def inputs: Set[String] = Set.empty
+}
+
+object InitialWorkDirFileGeneratorExpression {
+  type InitialWorkDirFileEvaluator = (IoFunctionSet, ParameterContext, ExpressionLib) => ErrorOr[WomValue]
+
+  /**
+    * Converts an InitialWorkDir.
+    *
+    * TODO: Review against the spec. Especially for Dirent values. For example:
+    *
+    * "If the value is an expression that evaluates to a Dirent object, this indicates that the File or Directory in
+    * entry should be added to the designated output directory with the name in entryname."
+    *
+    * - http://www.commonwl.org/v1.0/CommandLineTool.html#InitialWorkDirRequirement
+    * - http://www.commonwl.org/v1.0/CommandLineTool.html#Dirent
+    */
+  object InitialWorkDirFilePoly extends Poly1 {
+    implicit val caseExpressionDirent: Case.Aux[ExpressionDirent, InitialWorkDirFileEvaluator] = {
+      at { expressionDirent =>
+        (ioFunctionSet, parameterContext, expressionLib) => {
+          val entryEvaluation = ExpressionEvaluator.eval(expressionDirent.entry, parameterContext, expressionLib)
+          entryEvaluation flatMap {
+            case womFile: WomFile =>
+              val errorOrEntryName: ErrorOr[String] = expressionDirent.entryname match {
+                case Some(actualEntryName) => actualEntryName.fold(EntryNamePoly).apply(parameterContext, expressionLib)
+                case None => womFile.value.split('/').last.valid
+              }
+              errorOrEntryName flatMap { entryName =>
+                validate(Await.result(ioFunctionSet.copyFile(womFile.value, entryName), Duration.Inf))
+              }
+            case other => for {
+              coerced <- WomStringType.coerceRawValue(other).toErrorOr
+              contentString = coerced.asInstanceOf[WomString].value
+              // We force the entryname to be specified, and then evaluate it:
+              entryNameStringOrExpression <- expressionDirent.entryname.toErrorOr(
+                "Invalid dirent: Entry was a string but no file name was supplied")
+              entryName <- entryNameStringOrExpression.fold(EntryNamePoly).apply(parameterContext, expressionLib)
+              writeFile = ioFunctionSet.writeFile(entryName, contentString)
+              writtenFile <- validate(Await.result(writeFile, Duration.Inf))
```

Comment:
Probably? This PR was primarily aimed at fixing "the docker". This section here is a Poly refactoring of the previous pattern-matching code aimed at removing the `???`. No existing conformance tests were injured or harmed during the copy/paste of the former code.

I will be the owner under the "last touched" rule, but I've left a comment above that points out that the Dirent logic _doesn't look quite right_ and to change as necessary.

Re: your specific comment, yep `writeFile` would fail on a directory. CWL Directories are each effectively a recursive `listing: Option[Array[FileOrDirectory]]`. If there's `Some()` listing, one is supposed to write the supplied listing. If the listing is `None` one is supposed to query the file-system to find out what files are under that path. Our expression code will decode into a `Directory` object, but we'll likely need more wiring for handling Dirent / directories-in-general for InitialWorkDir.